fix #13994: initialise inline ctx in lateEnter

bishabosha opened the pull request:
In `lateEnter` the context is captured in the namer by the inline body annotation in `completeInCreationContext`. So we initialise the context for inlining before calling `lateEnter`.

I've not figured out a way to replicate this in sbt scripted tests, but you can experiment by copying the sources from #13994 to `library/src-bootstrapped`

fixes #13994

bishabosha commented:
mostly before this can be merged we need to add a test case that can reproduce the original issue - sbt scripted tests do not appear to work for this, I have been told that the cmdTests can be useful here

bishabosha commented:
@nicolasstucki I have added a case to bootstrapCmdTest - I would appreciate a better way to access the compiler version if possible

bishabosha commented:
This is ready to review - the test case reproduces the issue in #13994, requiring the fixes to `Run` to succeed

odersky reviewed (changes requested):
I think we need some changes. Detailed review will follow

odersky commented on compiler/src/dotty/tools/dotc/Run.scala:
```diff
@@ -303,17 +303,28 @@ class Run(comp: Compiler, ictx: Context) extends ImplicitRunInfo with Constraint
+            val phase = new transform.SetRootTree()
+            phase.run
+          if typeCheck then
+            val typerCtx: Context =
```
Simpler:
```scala
atPhase(Phase.typerPhase) {
  if compiling then finalizeActions += (() => typeCheckUnit())
  else typeCheckUnit()
}
```

odersky commented on compiler/src/dotty/tools/dotc/Run.scala:
```diff
@@ -303,17 +303,28 @@ class Run(comp: Compiler, ictx: Context) extends ImplicitRunInfo with Constraint
+        val namerCtx =
+          // inline body annotations are set in namer, capturing the current context
+          // we need to prepare the context for inlining.
+          if unit.isJava then ctx else PrepareInlineable.initContext(ctx)
```
I don't think we need the `isJava` distinction. So this could be
```scala
enterTrees()(using PrepareInlineable.initContext(ctx))
```

odersky reviewed (changes requested):
Maybe evaluate whether some of this can be moved to Namer and be merged with `lateEnter` there. It's important to keep driver classes like `Run` clean and not burden them with implementation detail.

bishabosha commented:
@odersky  I have moved most of the implementation of `lateCompile` to namer, save for a callback implemented in Run

smarter commented on compiler/src/dotty/tools/dotc/typer/Namer.scala:
```diff
@@ -708,15 +710,44 @@ class Namer { typer: Typer =>
+      // we need to prepare the context for inlining.
+      lateEnter()
+      typeCheckCB { () =>
+        atPhase(Phases.typerPhase) {
```
Aren't we already in a Context at phase typer here?

bishabosha replied:
nope, its at parser phase if you don't set this

bishabosha replied:
but before parser and typer used to be the same phase, so maybe that's what you're thinking

bishabosha replied:
maybe typer phase should be set before entering - the bug I came across was that implicit search from an inline method would not work

smarter replied:
yeah it should all be at typer phase.

bishabosha replied:
I reordered it a bit

smarter commented on compiler/src/dotty/tools/dotc/Run.scala:
```diff
@@ -302,20 +300,16 @@ class Run(comp: Compiler, ictx: Context) extends ImplicitRunInfo with Constraint
+        )
+
+      inContext(unitCtx) {
+        process()
       }
-      process()(using unitCtx)
     }
```
Previous version seemed fine?

bishabosha replied:
reverted

smarter commented on project/scripts/bootstrapCmdTests:
```diff
@@ -83,3 +83,12 @@ clear_out "$OUT"
 echo "Bug12973().check" | TERM=dumb ./bin/scala -cp "$OUT/out.jar" > "$tmp" 2>&1
 grep -qe "Bug12973 is fixed" "$tmp"
+
+echo "testing -sourcepath with inlining"
```
Add a comment explaining what this is testing exactly / why it can't be tested with an sbt scripted test?

smarter commented on project/scripts/bootstrapCmdTests:
```diff
@@ -83,3 +83,12 @@ clear_out "$OUT"
+
+echo "testing -sourcepath with inlining"
+cwd=$(pwd)
+java_prop="-Dpack.version.file=$cwd/dist/target/pack/VERSION"
+sbt_test_command=";clean;prepareSources;compile;copyChanges;compile"
```
isn't the property used here to do something equivalent to:
```suggestion
sbt_test_command="++$(cat dist/target/pack/VERSION)!;clean;prepareSources;compile;copyChanges;compile"
```

bishabosha replied:
I would like to do that but VERSION is a file with three properties inside

bishabosha replied:
found a solution without using java properties

bishabosha commented:
@smarter I added a comment to the cmd tests, I have also found a way to parse the scala version in bash so I deleted the changes related to loading scala version from java properties.

bishabosha commented:
@odersky would you like to check this again?

odersky reviewed (approved):
I think that way it is structured better than before.
